src/main/java/htsjdk/samtools/cram/structure/block/Block.java

```diff
@@ -0,0 +1,294 @@
+    }
+
+    public final BlockCompressionMethod getMethod() {
+        return method;
+    }
+
+    /**
+     * Identifies whether this is a header or data block, and which kind
+     * @return the CRAM content type of the block
+     */
+    public final BlockContentType getContentType() {
+        return contentType;
+    }
+
+    /**
+     * Return the External Content ID for this block.
+     *
+     * Only ExternalBlocks have a meaningful Content ID, so that class overrides this method.
+     *
+     * @return the External Content ID, or NO_CONTENT_ID
+     */
+    public int getContentId() {
+        return NO_CONTENT_ID;
+    }
+
+    public final byte[] getUncompressedContent() {
+        return ExternalCompression.uncompress(method, compressedContent);
+    }
+
+    /**
+     * The size of the uncompressed content in bytes.
+     */
+    public int getUncompressedContentSize() {
+        return getUncompressedContent().length;
+    }
+
+    /**
+     * Return the compressed block content
+     */
+
+    public final byte[] getCompressedContent() {
+        return compressedContent;
+    }
+
+
+    public final int getCompressedContentSize() {
+        return compressedContent.length;
+    }
+
+    /**
+     * Deserialize the Block from the {@link InputStream}. The reading is parameterized by the major CRAM version number.
+     *
+     * @param major CRAM version major number
+     * @param inputStream    input stream to read the block from
+     * @return a subtype of {@link Block} object with fields and content from the input stream
+     */
+    public static Block read(final int major, InputStream inputStream) {
+        final boolean v3OrHigher = major >= CramVersions.CRAM_v3.major;
+        if (v3OrHigher) {
+            inputStream = new CRC32InputStream(inputStream);
+        }
+
+        try {
+            final BlockCompressionMethod method = BlockCompressionMethod.byId(inputStream.read());
+            final BlockContentType type = BlockContentType.byId(inputStream.read());
+            final int contentId = ITF8.readUnsignedITF8(inputStream);
+            final int compressedSize = ITF8.readUnsignedITF8(inputStream);
+            final int rawSize = ITF8.readUnsignedITF8(inputStream);
+
+            final byte[] compressedContent = new byte[compressedSize];
+            InputStreamUtils.readFully(inputStream, compressedContent, 0, compressedSize);
+            if (v3OrHigher) {
+                final int actualChecksum = ((CRC32InputStream) inputStream).getCRC32();
+                final int checksum = CramInt.readInt32(inputStream);
+                if (checksum != actualChecksum) {
+                    throw new RuntimeException(String.format("Block CRC32 mismatch: %04x vs %04x", checksum, actualChecksum));
+                }
+            }
+
+            // TODO: is this check worthwhile?  it may be expensive.
+            final byte[] uncompressedContent = ExternalCompression.uncompress(method, compressedContent);
```

Comment:
The validation seems like a good idea, but it does seem like it would be expensive to do speculatively. Maybe we should do it lazily on "un-compression" instead. We'd have to retain the original size from the stream, but thats cheap.

Reply:
good idea, thanks